File: built-in plugins dsp/mb-reverb-shimmer.cpp
/**
 * MB Shimmer Reverb
 * Category : effect
 * Type     : reverb
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Pitched shimmer reverb
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_REVERB_SHIMMER_H
#define MB_REVERB_SHIMMER_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbReverbShimmer : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-reverb-shimmer";
    static constexpr const char* PLUGIN_NAME    = "MB Shimmer Reverb";
    static constexpr const char* PLUGIN_TYPE    = "reverb";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float shimmer = 0.6f;  // range [0, 1]
    float decay = 4.0f;  // range [1, 15]
    float mix = 0.4f;  // range [0, 1]
    };

    MbReverbShimmer() = default;
    ~MbReverbShimmer() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.shimmer = std::clamp(params.shimmer, 0f, 1f);
        params.decay = std::clamp(params.decay, 1f, 15f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Shimmer Reverb
        return input;
    }
};

#endif // MB_REVERB_SHIMMER_H

File: built-in plugins dsp/mb-reverb-shimmer.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbReverbShimmerPlugin: PluginDefinition = { id: 'mb-reverb-shimmer', slug: 'mb-reverb-shimmer', name: 'MB Shimmer Reverb', category: 'effect', type: 'reverb', version: '1.0.0', description: 'Pitched shimmer reverb', author: 'Max Booster', grade: 'A', parameters: [{ id: 'shimmer', name: 'Shimmer', type: 'float', defaultValue: 0.6, minValue: 0, maxValue: 1, automatable: true }, { id: 'decay', name: 'Decay', type: 'float', defaultValue: 4.0, minValue: 1, maxValue: 15, automatable: true }, { id: 'mix', name: 'Mix', type: 'float', defaultValue: 0.4, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { shimmer: 0.6, decay: 4.0, mix: 0.4 } };

export default MbReverbShimmerPlugin;
